lib/ash/resource/change/builtins.ex

```diff
@@ -46,20 +65,40 @@ defmodule Ash.Resource.Change.Builtins do
   @doc """
   Sets the attribute to the value provided.
 
-  If a zero argument function is provided, it is called to determine the value.
+  If a zero-argument function is provided, it is called to determine the value.
+  If a one-argument function is provided, it is called with the previous value to determine the value.
 
-    Use `arg(:argument_name)` to use the value of the given argument. If the argument is not supplied then nothing happens.
+  Use `arg(:argument_name)` to use the value of the given argument. If the argument is not supplied then nothing happens.
+
+  ## Options
+
+  #{Spark.OptionsHelpers.docs(Keyword.drop(@set_attribute_opts, [:attribute, :value]))}
 
   ## Examples
 
       change set_attribute(:active, false)
       change set_attribute(:opened_at, &DateTime.utc_now/0)
+      change set_attribute(:count, &(&1 + 1))
       change set_attribute(:status, arg(:status))
+      change set_attribute(:encrypted_data, arg(:data), set_when_nil?: false)
   """
-  @spec set_attribute(relationship :: atom, (() -> term) | {:_arg, :status} | term()) ::
+  def set_attribute_opts do
+    @set_attribute_opts
+  end
+
+  @spec set_attribute(
+          relationship :: atom,
+          (() -> term) | {:_arg, :status} | term(),
+          opts :: Keyword.t()
+        ) ::
           Ash.Resource.Change.ref()
-  def set_attribute(attribute, value) do
-    {Ash.Resource.Change.SetAttribute, attribute: attribute, value: value}
+  def set_attribute(attribute, value, opts \\ []) do
```

Comment:
to do this, you'll have to make this into a macro and do similar things to this:
https://github.com/ash-project/ash/blob/main/lib/ash/resource/change/builtins.ex#L240